### ISSUE TITLE
training of fullBGM 

### ISSUE THREAD

**Issue by wenxixi** [MCG-NJU/BCN#5]
Hi, Thanks for sharing the great work. 

I am training fullBGM() using your code. I didn't change your code but only added parts to print loss, precision, and recall for both train and val. This is what I got for dataset 50salads. 
![image](https://user-images.githubusercontent.com/42296651/114261410-f8d62180-9a0c-11eb-95d0-0abab29f2386.png)

Basically, val loss keeps increasing and val recall keeps decreasing from the start of training. Is this what should happen? Did I do anything wrong? The increasing loss seems to mean that the model does not learn at all.

It is written in your paper that you select the BGM with the best f1 score. May I ask what that best f1 score is? Based on my training curves so far, it seems that the best f1 score would still be quite a low value. 

Thank you.

**Comment by zhenzhiwang** [comment]
Thanks for your interest in our work. First of all, we want to clarify that our contribution lies in the LBP component, and the fullBGM code borrows directly from the PyTorch implementation of BSN (https://github.com/wzmsltw/BSN-boundary-sensitive-network.pytorch) and the only difference between BSN and ours is the hyper-parameters and training/testing datasets. As for the increasing val loss and decreasing val recall, we do not record this information in our experiments therefore we assume that the provided tensorboardX record is correct. We will try to explain it based on the curves:
1) Due to the successful dense boundary classification performance in BSN, we select it to classify action boundaries in our dataset, which should be effective in principle because of the similar data domain between action segmentation datasets, e.g., 50salads, and action detection datasets, e.g., activitynet. Therefore, if the val loss/recall is weird, maybe we should check the behavior of the BSN model on the activitynet dataset first. 
2) Furthermore, the groundtruth of action boundaries is manually constructed (a fixed ratio of action proposal near boundaries), therefore it may diverge from the final goal (segmentation result) a lot, where we may not care about the performance of this middle result, i.e., boundary classification result and its precision/recall metrics may not reflect the real properties we hope BGM provides. As discussed in our paper, the false positives have little influence on our final results because they only reduce the weights in the neighborhood, which is similar to smaller window sizes. However, we want the false negatives to be reduced because we need the action boundaries to separate different actions. As shown in the curves, the precision in the val set is increasing, therefore it is consistent with our expectations. Besides, even when the precision is low, i.e., there are no high activations in the regions that should be boundaries (which are really small regions), chances are that some neighborhood regions will have high activations, which leads to a shift of predicted action boundaries. In the calculation of precision/recall, it will not count as true predictions, yet it still provides useful information for refining segmentation results. In our view, the p/r metrics are just a value for choosing models for the next training stage, rather than a meaningful metric for this task.
3) In our experiments, we actually print the activation sequences and manually compare them to the ground truth boundaries. Based on the manual check on some (about 30) samples, we believe that the BGM module learns useful information, although there are some false positives and high activations with shifted locations.
4) Yes, the F1 score is a little low. Although the value differs in the three datasets, a typical value is about 0.4~0.5.  However, it works both in our task from the result of our experiments and in action detection task from the result of BSN model.
5) Although we didn't change the setting of loss in the original BSN implementation, we think that the highly unbalanced positive/negative samples may cause some problems. Maybe some adjustments such as Focal loss can be made here. We encourage you to try it both in our model and the original BSN model to see if there are problems in boundary classification modules.

If you have further questions, please feel free to add comments here.

**Comment by wenxixi** [comment]
Thank you so much for your explanations! I think I can understand your points here. I will follow some of  your suggestions to experiment and investigate a bit more. 

Thanks again!
